### PR TITLE
initial implementation of signup

### PR DESCRIPTION
**Pull request by phritz**
still a number of TODOs (see signup.go)

not hooked up for vercel yet but can run locally with `./diffs --db=/tmp/foo --account-db=/tmp/bar serve --signup-template-dir=serve/signup`

@arv fyi

progress towards https://github.com/rocicorp/repc/issues/269

### CODE REVIEW

File: account/db.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	ds datas.Dataset
+
+	mu   sync.Mutex
+	head Commit
+}
+
+// Commit is the git-like commit structure noms uses to store values.
+// The account database keeps a single Commit at the head of its dataset
+// as opposed to a chain of Commits representing changes. Noms requires
+// a Commit structure like this.
```
Comment (aboodman) [inline account/db.go:25]:
```suggestion
// Commit is the Git-like commit structure Noms uses to store values.
// The account database keeps a single Commit at the head of its dataset
// containing all current entries.
```

This is the way all Noms databases work, so it's a little confusing to have the "as opposed" bit.

File: account/db.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	defer r.lock()()
+	err := r.initLocked()
+	if err != nil {
+		return nil, err
+	}
+
+	return &r, nil
+}
+
+// ASIDs are issued in a separate range from regular accounts. See RFC.
```
Comment (aboodman) [inline account/db.go:62]:
```suggestion
// ASIDs are issued in a separate range from regular accounts.
// See RFC: https://github.com/rocicorp/repc/issues/269
```

File: account/db.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+// The account database keeps a single Commit at the head of its dataset
+// as opposed to a chain of Commits representing changes. Noms requires
+// a Commit structure like this.
+type Commit struct {
+	// Parents and Meta are unused.
+	Parents []types.Ref `noms:",set"`
+	Meta    struct {
+	}
+
+	Value Accounts
```
Comment (aboodman) [inline account/db.go:32]:
```suggestion
Value Accounts

// Store other fields Noms uses internally so they get roundtripped correctly.
Original noms.Struct `noms:",original"`
```

File: account/db.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	var head Commit
+	err := marshal.Unmarshal(db.ds.Head(), &head)
+	if err != nil {
+		return err
+	}
+	// Noms roundtrips empty maps as nil, so ensure we have a map.
+	if head.Value.AutoSignup == nil {
+		head.Value.AutoSignup = make(map[uint32]ASAccount)
```
Comment (aboodman) [inline account/db.go:81]:
```suggestion
		head.Value.AutoSignup = map[uint32]ASAccount{}
```

File: serve/service.go [hunk 5]
```diff
@@ -28,7 +28,7 @@ var (
 	pathRegex = regexp.MustCompile(`^\/([\w-]+)\/([\w-]+)\/([\w-]+)\/?$`)
 )
 
-// Service is a running instance of the Replicant service.
+// Service is an instance of the Replicache Diffserver services.
```
Comment (aboodman) [inline serve/service.go:31]:
```suggestion
// Service is an instance of the Replicache DiffServer service.
```

File: serve/signup/get.html [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+    <form method="POST" action="/signup">
+        <label>Name:</label> <input type="text" name="{{.Name}}"><br>
+        <label>Email:</label> <input type="text" name="{{.Email}}"><br>
+        <input type="submit">
+    </form>
+
+</body>
+
+</html>
```
Comment (aboodman) [inline serve/signup/get.html:27]:
```suggestion
</html>

```

File: serve/signup/get.html [hunk 7]
```diff
@@ -0,0 +1,27 @@
+<html>
```
Comment (aboodman) [inline serve/signup/get.html:1]:
```suggestion
<!doctype html>
<html>
```

File: serve/signup/get.html [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    <h2>Replicache Account Signup</h2>
+
+    <p>Please fill out the form below to generate an Account ID for Replicache.<br>
+        You need to include your Account ID in the <i>diffServerAuth</i> field when<br>
+        you <a href="https://github.com/rocicorp/replicache-sdk-js#%EF%B8%8F-instantiate">instantiate Replicache</a>
+        in your JavaScript application.
+
+    <p>Note that while your Account ID will be useful for <strong>evaluating</strong> Replicache,<br>
+        it maybe subject to certain limits that make it unsuitable for large-scale production use. <br>
+        Please contact us by TODO in order to have your account upgraded for produdction use.<br><br>
```
Comment (arv) [inline serve/signup/get.html:17]: typo


produdction -> production

File: serve/signup/get.html [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+    <form method="POST" action="/signup">
+        <label>Name:</label> <input type="text" name="{{.Name}}"><br>
```
Comment (arv) [inline serve/signup/get.html:20]:
```html
<label>Name: <input type="text" name="{{.Name}}"></label><br>
```

**Review by arv — COMMENTED** [review]
Nice

File: account/db.go (hunk 3 above)
Comment (phritz) [reply account/db.go:32]: can you splain this one to me?

File: serve/signup/get.html (hunk 6 above)
Comment (phritz) [reply serve/signup/get.html:27]: my editor keeps trimming the whitespace

**Comment by phritz** [comment]
Thanks. Addressed in https://github.com/rocicorp/diff-server/pull/84 except for original, which i'm waiting to hear more about.

File: account/db.go (hunk 3 above)
Comment (aboodman) [reply account/db.go:32]: Nevermind. I misunderstood.

**Pull request merged**